thermostat functionality WIP

daenny opened the pull request:
This enables the Alexa Thermostat functionality for climate components in Homeassistant

auchter commented on haaska.py:
```diff
@@ -188,8 +193,12 @@ def response_wrapper(ha, payload):
             try:
-                func(ha, payload)
-                r['payload'] = {'success': True}
+                res = func(ha, payload)
```
Reading the docs again, apparently it's no longer necessary to include {'success': True} in the response payload. So maybe change these few lines to:
`r['payload'] = func(ha, payload) or {}`

daenny replied:
I did not know that this assignment works, Python is great ;).

daenny commented on haaska.py:
```diff
@@ -328,6 +391,29 @@ def set_percentage(self, val):
+        target_temp = state['attributes']['temperature']
+        heating_mode = 'AUTO'
+        if state['attributes']['away_mode'] == 'on':
+            heating_mode = 'AWAY'
```
Yeah I was wondering about that. 
You can have some deviceMode "AWAY" (https://developer.amazon.com/public/solutions/alexa/alexa-skills-kit/docs/smart-home-skill-api-reference#notsupportedincurrentmodeerror)
It does not seem to matter anyway, since the value is actually not used at all. But we maybe should keep it according to the docs.

daenny commented on haaska.py:
```diff
@@ -328,6 +391,29 @@ def set_percentage(self, val):
+    def turn_off(self):
+        self._call_service('climate/set_away_mode', {'away_mode': True})
+
+    def get_state(self):
```
I agree, get state should move up to Entity. In the climate entity I would then maybe include get_min_temp() and get_max_temp() and get_mode(), to make it cleaner in the handlers for the temperature requests. 
There is also some duplication for the range check, for setTemperature and for adjTemperature. Maybe it's better to just add a "is_valid_temp()" to the ClimateEntity.

daenny commented:
About the changes of modes: How does Home Assistant handle this?
My thermostat has only modes: comfort, away, long-away and fixed temperature. 
It was not supported yet, so I implemented it myself. I automatically change the mode to comfort mode when I change the temperature. The thermostat automatically switches back to away after a certain amount of time. So for me setting the temperature always works, no matter which mode. 
How is this handled with the Nest? Can you set a temperature via the GUI while being in off mode? 

In general, there are two errors that you could throw to deal with this:
https://developer.amazon.com/public/solutions/alexa/alexa-skills-kit/docs/smart-home-skill-api-reference#notsupportedincurrentmodeerror
Or:
https://developer.amazon.com/public/solutions/alexa/alexa-skills-kit/docs/smart-home-skill-api-reference#unwillingtosetvalueerror

daenny commented on haaska.py:
```diff
@@ -258,6 +267,60 @@ def handle_decrement_percentage(ha, payload):
+@handle('DecrementTargetTemperatureRequest')
+@control_response('DecrementTargetTemperatureConfirmation')
+def handle_decrement_percentage(ha, payload):
+    handle_temperature_adj(ha, payload, operator.sub)
```
I just realized, I am missing the return here actually

sibbl commented:
Just got my Echo yesterday, but haaska is now running fine with my Danfoss LC-13 thermostat controls. However, I had to adapt it as the Z-Wave thermostats don't support modes or specifially the `away_mode` property. Is it possible to make it more generic?

1. The main problem is that there's no `away_mode` property in the `attributes` object, which first of all made `get_state` always fail for me. 
2. I extended the `turn_on` and `turn_off` methods to call `get_state` before triggering some action. If there's no `away_mode`, then it doesn't call `set_away_mode` but `set_temperature` with 10°C (off) and 24°C (on) instead. Would it make sense to implement this feature generally with both values being configurable?
3. Some quick question about Alexa: when setting the temperature, it always tells me the current target temperature and in which mode it is. I didn't find a way to skip the last information as it's always "auto" for me. Do you know how one could fix this as well?

Finally, let me thank you for the project and the hard work :)

daenny commented:
At the moment I am a bit busy with other projects, i.e. work, so I do not know if I have time to do the changes that I proposed. 
@sibbl: 
1. You can change line 405 to: 
` if 'away_mode' in state['attributes'] and state['attributes']['away_mode'] == 'on':`
But actually this is not (yet?) used by Alexa. So maybe just remove it completely. 
2. Yes, I agree, it could probably go into the configuration file. Something like:
```
climate_on: [{set_temperature: 24}]
climate_off: [{set_away_mode: True}, {set_temperature: 14}]
```
I would make it an array, since maybe multiple things could be called.
3. You can just change the response to always return HEAT as mode, but then Alexa says:
"Set thermostat to heat to x degrees" (which sounds a bit mumbled to me) instead of "Thermostat is aiming for x degrees, thermostat is in auto mode". I liked latter better, even though it also says the unnecessary mode. Just try it out. 
But this should probably be also go in the config file. I.e. something like: 
`climate_return_mode: AUTO
`

keatontaylor commented:
I would love to see this, especially as a means to hack reporting of sensor values to Alexa in addition to controlling thermostats.

auchter commented:
Closing since this was implemented by #42
